### PR TITLE
Export source classes to rspec-support

### PR DESCRIPTION
**Pull request by yujinakayama**
For https://github.com/rspec/rspec-expectations/pull/987.

This corresponds to https://github.com/rspec/rspec-support/pull/315.

### CODE REVIEW

**Comment by myronmarston** [comment]
@yujinakayama what is the reason to move this to rspec-support?

Also, are you hoping to get this in RSpec 3.6?  I believe @samphippen will release that today (once #2419 gets merged).

**Comment by yujinakayama** [comment]
This is for https://github.com/rspec/rspec-expectations/pull/987.

> Also, are you hoping to get this in RSpec 3.6? I believe @samphippen will release that today (once #2419 gets merged).

I don't expect so but it would be nice if it could.

**Comment by myronmarston** [comment]
> I don't expect so but it would be nice if it could.

Makes sense.  The feature in rspec/rspec-expectations#987 is really nice, and things look very close to ready to merge.  I'd lean towards trying to get this in before we release if we can get this merged in the next 12-24 hours, but not letting it hold up the release for multiple days.  Ultimately, I'll leave that decision up to @samphippen, though.

I'll go review your PRs now.

File: Gemfile [hunk 1]
```diff
@@ -12,6 +12,8 @@ branch = File.read(File.expand_path("../maintenance-branch", __FILE__)).chomp
   end
 end
 
+gem 'rspec-support', :git => "https://github.com/rspec/rspec-support.git", :branch => 'import-source-classes-from-rspec-core'
```
Comment (myronmarston) [inline Gemfile]: Let's not forget to change this back before we merge.

File: lib/rspec/core/world.rb [hunk 2]
```diff
@@ -128,11 +128,18 @@ def reporter
       end
 
       # @private
-      def source_cache
-        @source_cache ||= begin
-          RSpec::Support.require_rspec_core "source"
-          Source::Cache.new(@configuration)
+      def source_from_file(path)
+        unless instance_variable_defined?(:@sources_by_path)
```
Comment (myronmarston) [inline lib/rspec/core/world.rb]: Any reason to favor this over `unless defined?(@sources_by_path)`?

File: script/clone_all_rspec_repos [hunk 3]
```diff
@@ -15,7 +15,7 @@ if is_mri; then
   clone_repo "rspec-rails"
 
   if rspec_support_compatible; then
-    clone_repo "rspec-support"
+    MAINTENANCE_BRANCH=import-source-classes-from-rspec-core clone_repo "rspec-support"
```
Comment (myronmarston) [inline script/clone_all_rspec_repos]: Don't forget to change this back, either.

File: lib/rspec/core/world.rb [hunk 4]
```diff
@@ -128,11 +128,18 @@ def reporter
...
+          RSpec::Support.require_rspec_support 'source'
+          @sources_by_path = {}
```
Comment (myronmarston) [inline lib/rspec/core/world.rb]: Do we need to clear `@sources_by_path` (e.g. `@sources_by_path.clear`) when `world.reset` is called?  I'm thinking maybe we do....

Comment (yujinakayama) [inline lib/rspec/core/world.rb]: Done

File: lib/rspec/core/world.rb (hunk 2 above)
Comment (yujinakayama) [inline lib/rspec/core/world.rb]: Changed to `defined?(@sources_by_path)`.

**Pull request merged**